src/goto-instrument/contracts/assigns.cpp

```diff
@@ -24,49 +24,30 @@ assigns_clause_targett::assigns_clause_targett(
   code_contractst &contract,
   messaget &log_parameter,
   const irep_idt &function_id)
-  : pointer_object(pointer_for(object)),
-    contract(contract),
+  : contract(contract),
     init_block(),
     log(log_parameter),
-    target(typet()),
+    target(pointer_for(object)),
     target_id(object.id())
 {
   INVARIANT(
-    pointer_object.type().id() == ID_pointer,
+    target.type().id() == ID_pointer,
     "Assigns clause targets should be stored as pointer expressions.");
-  const symbolt &function_symbol = contract.ns.lookup(function_id);
-
-  // Declare a new symbol to stand in for the reference
-  symbolt standin_symbol = contract.new_tmp_symbol(
-    pointer_object.type(),
-    function_symbol.location,
-    function_symbol.mode);
-
-  target = standin_symbol.symbol_expr();
-
-  // Build standin variable initialization block
-  init_block.add(goto_programt::make_decl(target, function_symbol.location));
-  init_block.add(goto_programt::make_assignment(
-    code_assignt(target, pointer_object), function_symbol.location));
 }
 
 assigns_clause_targett::~assigns_clause_targett()
 {
 }
 
-std::vector<symbol_exprt> assigns_clause_targett::temporary_declarations() const
-{
-  std::vector<symbol_exprt> result;
-  result.push_back(target);
-  return result;
-}
-
 exprt assigns_clause_targett::alias_expression(const exprt &lhs)
 {
   exprt::operandst condition;
   exprt lhs_ptr = (lhs.id() == ID_address_of) ? to_address_of_expr(lhs).object()
                                               : pointer_for(lhs);
 
+  // NULL targets aren't writable
+  condition.push_back(not_exprt(null_pointer(target)));
```

Comment:
Should this check be more generally about validity, not just non-`NULL`-ness? Like the w_ok check we use in `compatible_expression`?

Reply:
We could add a check similar to
```cpp
condition.push_back(w_ok_exprt(target, from_integer(object_size(target), unsigned_int_type())));
```
How is that better than this nullness check? For me, it looks a bit redundant, because I get the size of the object, then I check if the object is writable up to its given size.

Reply:
> because I get the size of the object, then I check if the object is writable up to its given size.

I might have missed this. If you do this, then yes it's redundant. Could you please point me to the specific lines that implement this?

> How is that better than this nullness check?

The idea is to use `w_ok` to make sure it's writable.

Reply:
> Could you please point me to the specific lines that implement this?

I meant, I could replace the line 49 with this:
```c
condition.push_back(w_ok_exprt(target, from_integer(object_size(target), unsigned_int_type())));
```

It's not part of our alias check yet.

Reply:
Implemented as suggested.